### PR TITLE
First implementation of backlink testing.

### PR DESCRIPTION
**Pull request by kalaspuffar**
Trying to solve #478

### CODE REVIEW

File: src/main/resources/xml/schema/2020-1/nordic2020-1.opf-and-html.sch [hunk 1]
```diff
@@ -188,4 +188,30 @@
         </rule>
     </pattern>
 
+
+    <pattern id="nordic_opf_and_html_30_a">
+        <title>Rule 30a</title>
+        <p>Each noteref must have a backlink</p>
+        <rule context="html:*[tokenize(@epub:type,'\s+')=('noteref')]">
```
Comment (josteinaj) [inline src/main/resources/xml/schema/2020-1/nordic2020-1.opf-and-html.sch:195]: I think noterefs is always `<a>`?

File: src/main/resources/xml/schema/2020-1/nordic2020-1.opf-and-html.sch [hunk 2]
```diff
@@ -188,4 +188,30 @@
...
+            <let name="context" value="concat('(&lt;', name(), string-join(for $a in (@*) return concat(' ', $a/name(), '=&quot;', $a, '&quot;'), ''), '&gt;)')"/>
+            <!-- this is the multi-HTML version of the rule; the single-HTML version of this rule is in nordic2020-1.sch -->
+            <assert test="count(//html:a[tokenize(@role,'\s+')='doc-backlink' and substring-after(@href,'#') = current()/@id]) &gt;= 1">[nordic_opf_and_html_30a] The <value-of
```
Comment (josteinaj) [inline src/main/resources/xml/schema/2020-1/nordic2020-1.opf-and-html.sch:198]: Can there be multiple backlinks or should this check for exactly one match?

File: src/main/resources/xml/schema/2020-1/nordic2020-1.opf-and-html.sch [hunk 3]
```diff
@@ -188,4 +188,30 @@
...
+        </rule>
+    </pattern>
+
+    <pattern id="nordic_opf_and_html_30_b">
+        <title>Rule 30b</title>
+        <p>Each backlink must reference a noteref</p>
+        <rule context="html:a[tokenize(@role,'\s+')='doc-backlink']">
+            <let name="context" value="concat('(&lt;', name(), string-join(for $a in (@*) return concat(' ', $a/name(), '=&quot;', $a, '&quot;'), ''), '&gt;)')"/>
+            <!-- this is the multi-HTML version of the rule; the single-HTML version of this rule is in nordic2020-1.sch -->
+            <assert test="count(//html:*[tokenize(@epub:type,'\s+')=('noteref') and @id = current()/substring-after(@href,'#')]) &gt;= 1">[nordic_opf_and_html_30b] The backlink
```
Comment (josteinaj) [inline src/main/resources/xml/schema/2020-1/nordic2020-1.opf-and-html.sch:211]: Here also, I think noterefs are aways `<a>`.

And there should probably be just one noteref matching each backlink?

@martinpub I suppose a book might potentially have multiple noterefs referencing the same note. How do we handle backlinks in such cases?

File: src/main/resources/xml/schema/2020-1/nordic2020-1.sch [hunk 4]
```diff
@@ -163,6 +163,16 @@
         </rule>
     </pattern>
 
+    <pattern id="epub_nordic_26_c">
+        <title>Rule 26c</title>
+        <p>Each note must have a backlink</p>
+        <rule context="html:*[tokenize(@epub:type, '\s+') = ('note', 'endnote', 'footnote')]">
+            <let name="context" value="concat('(&lt;', name(), string-join(for $a in (@*) return concat(' ', $a/name(), '=&quot;', $a, '&quot;'), ''), '&gt;)')"/>
+            <!-- this is the single-HTML version of the rule; the multi-HTML version of this rule is in nordic2020-1.opf-and-html.sch -->
```
Comment (josteinaj) [inline src/main/resources/xml/schema/2020-1/nordic2020-1.sch:171]: these single-HTML vs multi-HTML comments can probably be removed.

**Review by josteinaj — CHANGES_REQUESTED** [review]
Looks good, should work. But I have some comments/questions.

**Comment by kalaspuffar** [comment]
Hi @josteinaj 

Thank you for the review. I've updated the code after your suggestions.

Best regards
Daniel

**Comment by josteinaj** [comment]
Looks good!

@martinpub How should be handle backlinks in books with multiple noterefs referencing the same note? I assume that's not too uncommon?

File: src/main/resources/xml/schema/2020-1/nordic2020-1.opf-and-html.sch (hunk 3 above)
Comment (josteinaj) [reply src/main/resources/xml/schema/2020-1/nordic2020-1.opf-and-html.sch]: (resolving this, asked here instead: https://github.com/nlbdev/nordic-epub3-dtbook-migrator/pull/509#issuecomment-960876785)

**Review by josteinaj — APPROVED** [review]

**Comment by josteinaj** [comment]
I think we need to solve the case of multiple noterefs referencing the same note before we merge this.

**Comment by josteinaj** [comment]
The conclusion is that there must be exactly one note reference for each note (see https://github.com/nlbdev/epub3-guidelines-update/issues/150).

**Review by josteinaj — CHANGES_REQUESTED** [review]
We should also change the annotation rules:
- Every annotation reference must have exacly one annotation
- Every annotation must have exactly one annotation reference

Other than that, I think it looks good.

**Comment by kalaspuffar** [comment]
@josteinaj Made the requested text changes.

**Review by josteinaj — APPROVED** [review]

**Review by erikmtm — APPROVED** [review]

**Comment by josteinaj** [comment]
Merged into `master` in 15daadbc5fc213b9327fc0ddaa2df7c4603332cf

**Pull request closed without merging**